tests/timeslot_notification_test.php

```diff
@@ -0,0 +1,214 @@
+     */
+    private function create_valid_timeslot() {
+        $obid = $this->instance->id;
+        $data = self::VALID_DATA;
+        $data['observer_id'] = $this->observer->id;
+        $data['obs_id'] = $obid;
+
+        return \mod_observation\timeslot_manager::modify_time_slot($data);
+    }
+
+    /**
+     * Test basic CRUD operations for notifications
+     */
+    public function test_notification_crud() {
+        $this->setUser($this->observee);
+
+        // Create notification.
+        \mod_observation\timeslot_manager::create_notification($this->instance->id, $this->slot1id, $this->observee->id,
+            (object) self::NOTIFY_DATA);
+
+        $notifys = \mod_observation\timeslot_manager::get_users_notifications($this->instance->id, $this->observee->id);
+        $this->assertEquals(1, count($notifys));
+
+        // Delete notification.
+        \mod_observation\timeslot_manager::delete_notification($this->instance->id, $this->observee->id,
+            array_values($notifys)[0]->notification_id);
+
+        $notifys = \mod_observation\timeslot_manager::get_users_notifications($this->instance->id, $this->observee->id);
+        $this->assertEquals(0, count($notifys));
+    }
+
+    /**
+     * Test if user can delete a notification they didn't create
+     */
+    public function test_delete_not_owned() {
+        $this->setUser($this->observee);
+
+        // Create notification for observee 1.
+        \mod_observation\timeslot_manager::create_notification($this->instance->id, $this->slot1id, $this->observee->id,
+            (object) self::NOTIFY_DATA);
+        $notifys = \mod_observation\timeslot_manager::get_users_notifications($this->instance->id, $this->observee->id);
+        $this->assertEquals(1, count($notifys));
+
+        // Try and delete as observee 2 (not allowed).
+        $this->expectException('moodle_exception');
+        \mod_observation\timeslot_manager::delete_notification($this->instance->id, $this->observee2->id,
+            array_values($notifys)[0]->notification_id);
+    }
+
+    /**
+     * Tests if a user can create too many notifications than allowed
+     */
+    public function test_create_too_many() {
+        $this->setUser($this->observee);
+
+        // Create the allowed number.
+        for ($i = 0; $i < \mod_observation\timeslot_manager::MAX_NOTIFICATIONS; $i++) {
+            \mod_observation\timeslot_manager::create_notification($this->instance->id, $this->slot1id, $this->observee->id,
+                (object) self::NOTIFY_DATA);
+        }
+
+        // Try and create 1 more (error).
+        $this->expectException('moodle_exception');
+        \mod_observation\timeslot_manager::create_notification($this->instance->id, $this->slot1id, $this->observee->id,
+            (object) self::NOTIFY_DATA);
+    }
+
+    /**
+     * Tests invalid inputs for the interval field
+     */
+    public function test_invalid_interval() {
+        $invalidvars = [
```

Comment:
This looks like it might be nicely suited to a mini data provider https://docs.moodle.org/dev/Writing_PHPUnit_tests#Make_use_of_the_dataProvider_functionality

Reply:
The data providers look interesting, however, I think in this case because only a single property is changing each time, it would be a bit unnecessary.